cli/flox-rust-sdk/src/providers/buildenv.rs

```diff
@@ -158,20 +170,84 @@ impl BuildEnvNix {
     /// See the individual realisation functions for more details.
     // todo: return actual store paths built,
     // necessary when building manifest builds.
-    fn realise_lockfile(&self, lockfile: &Lockfile, system: &System) -> Result<(), BuildEnvError> {
+    fn realise_lockfile(
+        &self,
+        client: &impl ClientTrait,
+        lockfile: &Lockfile,
+        system: &System,
+    ) -> Result<(), BuildEnvError> {
         for package in lockfile.packages.iter() {
             if package.system() != system {
                 continue;
             }
             match package {
-                LockedPackage::Catalog(locked) => self.realise_nixpkgs(locked)?,
+                LockedPackage::Catalog(locked) => self.realise_nixpkgs(client, locked)?,
                 LockedPackage::Flake(locked) => self.realise_flakes(locked)?,
                 LockedPackage::StorePath(locked) => self.realise_store_path(locked)?,
             }
         }
         Ok(())
     }
 
+    /// Try to substitute a custom package by copying it from an associated store.
+    ///
+    /// Query the associated store(s) that contain the package from the catalog.
+    /// Then attempt to download the package closure from each store in order,
+    /// until successful.
+    /// Returns `true` if all outputs were found and downloaded, `false` otherwise.
+    fn try_substitute_custom_pkg(
+        &self,
+        client: &impl ClientTrait,
+        locked: &LockedPackageCatalog,
+    ) -> Result<bool, BuildEnvError> {
+        // TODO - The API call accepts multiple, so an optimization is to
+        // collect these for the whole lockfile ahead of time and ask for them
+        // all at once.
+        let paths: Vec<String> = locked.outputs.values().map(|s| s.to_string()).collect();
+        let store_locations = client
+            .get_store_info(paths)
+            .block_on()
+            .map_err(BuildEnvError::CatalogError)?;
+
+        // Try downloading each output from the store location provided.  If we
+        // are missing store info for any, we should return false.
+        // TODO - It is possible not _all_ are missing. For now, we'll just
+        // assume they are all missing, noting custom packages only have one
+        // output currently.  Also to note: if all the outputs were valid
+        // locally, we would not get here as that check happens before this is
+        // called within `realise_nixpkgs`.
+        let mut all_found = true;
+        'path_loop: for (path, locations) in store_locations.iter() {
+            // If there are no locations
+            for location in locations {
+                // nix copy
+                let mut copy_command = Command::new(&*NIX_BIN);
+                copy_command
+                    .arg("copy")
+                    .arg("--from")
+                    .arg(&location.url)
+                    .arg(path);
```

Comment:
suggestion(nb): log copy command when executed

```suggestion
                copy_command
                    .arg("copy")
                    .arg("--from")
                    .arg(&location.url)
                    .arg(path);
                    
                debug!(cmd=%copy_command.display(), %path, url=%location.url, "trying to download store path");
```

Reply:
i can fold that into the spinner pr since that will requre handling merge conflicts